API_accelerometer

aleball20 opened the pull request:
Created an API with 3 instruzions:

int accelerometer_init(void);								//initialize value int accelerometer_get_data_raw(accelerometer_data_t* data);	// get data in raw value int accelerometer_get_data_g(accelerometer_data_t* data_g);  // get data in g vale

accelerometer_data_t is a structur which contains the 3 directions: x,y,z

leonardobove committed:
Add software project with basic scheduler

aleball20 committed:
API_accelerometer

Created an API with 3 instruzions:

int accelerometer_init(void);								//initialize value
int accelerometer_get_data_raw(accelerometer_data_t* data);	// get data in raw value
int accelerometer_get_data_g(accelerometer_data_t* data_g);  // get data in g vale

accelerometer_data_t is a structur which contains the 3 directions: x,y,z

leonardobove committed:
Merge branch 'sys_tick' into accelerometer_spi_test

leonardobove committed:
Remove useless files

leonardobove commented:
@aleball20
Ti ho rimosso tutti i file inutili che vengono automaticamente creati dal progetto di Eclipse. Avevo aggiornato il .gitignore sul mio branch, mi ero dimenticato di passarvelo. Come vedi adesso la PR è molto più pulita e mi focalizzo sulle modifiche importanti che hai fatto. **IMPORTANTE**: fai un pull del branch prima di fare le modifiche che ti elenco di seguito.

La struttura dell'API in generale va bene (un .c e un .h), tuttavia solitamente si usano gli stessi nomi per il .c e per il .h, per capire al volo quali sono i file associati tra di loro: ti consiglio di dare a entrambi lo stesso nome.
Ottima l'idea di aggiungere un sorgente di test, per mostrare agli utenti come utilizzare la tua API. L'unica cosa è che bisogna stare attenti che non ci finisca nei MakeFile, altrimenti viene compilato e ci darà sicuramente errore (quando non vogliamo che venga compilato ovviamente)

leonardobove commented on software/API_accelerometer/Inc/api_accelerometer.h:
```diff
@@ -0,0 +1,43 @@
+#ifndef INC_API_ACCELEROMETER_H_
+#define INC_API_ACCELEROMETER_H_
+
+#include "altera_up_avalon_accelerometer_spi.h"
+#include <stdint.h>
```
E' buona pratica quella di mettere prima tutte le librerie standard (<>) e poi tutte quelle private (""), perché quelle private potrebbero essere fatte apposta per sovrascrivere alcuni define delle librerie standard e purtroppo il compilatore legge dall'alto verso il basso

leonardobove commented on software/API_accelerometer/Inc/api_accelerometer.h:
```diff
@@ -0,0 +1,43 @@
+} accelerometer_range_t;*/
+
+// **** lecture data *****
+typedef struct {
+    int x;
+    int y;
+    int z;
+} accelerometer_data_t;
```
Ottima l'idea del typedef. Tuttavia non mi piace troppissimo vedere `int` in un codice embedded... Che ne dici di metterci `uint32_t`? Il motivo per cui `int` è pericoloso lo conosci.

leonardobove commented on software/API_accelerometer/Inc/api_accelerometer.h:
```diff
@@ -0,0 +1,43 @@
+
+
+// ***** MACROS *****
+#define REG_DATA_FORMAT 0x31
```
Ok, buona pratica che non ci ha fatto vedere Baronti è di definire le macro costanti letterali usando lo specificatore del tipo di letterale e le parentesi (questo te lo spiegherò a voce, sostanzialmente diamo una mano al compilatore a non fare errori stupidi). Quindi sarebbe meglio scrivere:
```
#define REG_DATA_FORMAT (0x31u)
```

leonardobove commented on software/API_accelerometer/Inc/api_accelerometer.h:
```diff
@@ -0,0 +1,43 @@
+
+//**** API *****
+
+int accelerometer_init(void);								//initialize value
+//int accelerometer_set_range(accelerometer_range_t range);	// set range
+int accelerometer_get_data_raw(accelerometer_data_t* data);	// get data in raw value
+int accelerometer_get_data_g(accelerometer_data_t* data_g);  // get data in g vale
```
Please no `int`

leonardobove commented on software/API_accelerometer/Inc/config.h:
```diff
@@ -15,7 +15,6 @@
 #include <stdint.h>
 #include <stdio.h>
 #include "unistd.h" // per usleep()
-#include "altera_up_avalon_accelerometer_spi.h"
```
Il `config.h` non ti serve. Se queste librerie servono per far funzionare la tua API, includile tutte dentro `api_accelerometer.h`. `config.h` lo lasciamo per il main code

leonardobove commented on software/API_accelerometer/Scr/test_api.c:
```diff
@@ -0,0 +1,31 @@
+ *      Author: aless
+ */
+
+#include "../Inc/config.h"
```
Include tutto con `api_accelerometer.h`

leonardobove commented on software/API_accelerometer/Scr/accelerometer.c:
```diff
@@ -0,0 +1,55 @@
+//static accelerometer_range_t current_range = ACCELEROMETER_RANGE_2G; //Default init has 2G
+
+//**** type accelerometer ****
+static alt_up_accelerometer_spi_dev *accel_dev;
+
+int accelerometer_init(void){
+	accel_dev = alt_up_accelerometer_spi_open_dev("/dev/accelerometer_spi_0");
```
Mi ero dimenticato di commentare questa cosa: ha senso lasciare nascosto in quel modo il puntatore al gestore dell'accelerometer_spi? Adesso così è `static` senza bisogno di esserlo e non viene mai ritornato all'utente, nel caso in cui ne avesse bisogno per qualche motivo. Io ti direi di farlo ritornare dalla funzione `alt_up_accelerometer_spi_dev`. @aleball20

leonardobove commented on software/API_accelerometer/Inc/api_accelerometer.h:
```diff
@@ -0,0 +1,43 @@
+/*
+ * api_accelerometer.h
+ *
+ *  Created on: 17 apr 2025
+ *      Author: aless
+ */
```
Potresti aggiungere in questa zona di commenti alcune specifiche generiche che fanno comodo? Tipo che tipo di accelerometro è (il codice del modello), la risoluzione e la massima portata impostabile. @aleball20

leonardobove commented:
@aleball20 ma alla fine quindi hai fatto in modo tale che vengano ritornati i valori int32_t non convertiti provenienti direttamente dal sensore, giusto? E su quel problema sui floating point hai indagato su come risolvere?

aleball20 committed:
fixed all the tips (declaration and comements)

Ho fixato tutte le cose che mi hai detto,.
Ho rimosso il test_Accelerometer dal makefile. Adesso il compilatore mi da un errore, ma è solo dovuto al fatto che non trova il main.

Per quanto riguarda il discorso del floating point, si ho controllato. la CPU economy  non ha un divisor HW mentre la fast si.

La funzione int32_t accelerometer_get_data_g che  stampa i valori in [m/s^2] funziona comunque dato che l'unico valore di tipo float nel codice è una costante e quindi viene gestito via SW porbabilmente come un fixed point, come ci aveva insegnato Saletti.  Essendo però l'uscita di tipo int32_t, i valori decimali vengono tagliati (per intenderci, se tengo ferma la scheda sul tavolo si deve che lungo Z il valore è 9 valore intero e non 9.81).

Nel complesso secondo me possiamo tenere il processore economy, dato che la funzione int32_t accelerometer_get_data_g l'ho inserita solo a scopo di test, utile per una stampa a video, mentre suggerisco di usare per il progetto la funz. accelerometer_get_data_raw che stampa solo valori interi da -255 a 255 esattamente come vengono forniti dall'accelerometro, senza rischiare che il software arrotondi la misura effettuata.

leonardobove commented:
@aleball20 
![image](https://github.com/user-attachments/assets/7ae4b352-cddf-430e-8bc3-3f63a06577ee)
C'è un po' di casino? Magari volevi rimuovere i file vecchi?
MI pare che non ci siano differenze tra `accelerometer.c` e `test_api.c`. E perché i file .h e .c della libreria non sono nella stessa cartella? Cioè andava bene prima avere `API_accelerometer` con dentro `Inc` e `Src` (va bene anche non usare Inc e Src in questo caso e mettere tutto dentro `API_accelerometer`  direttamente)

leonardobove committed:
Fix library structure
